### DISCUSSION TITLE
Not initialized x509_st->skid. At what point is the x509_st->skid field initialized?

### DISCUSSION THREAD

**Discussion by SergeyKiwi** [openssl/openssl#22398] [Q&A] [answered]
I get error when OpenSSL look for a CRL issuer with extended_crl flag. I debugged the code and found that for some certificates contain not empty skid, while others have this field equal to 0. Therefore X509_check_akid returns True. 
```
for (i = 0; i < sk_X509_num(ctx->untrusted); i++) {
    crl_issuer = sk_X509_value(ctx->untrusted, i);
    if (X509_NAME_cmp(X509_get_subject_name(crl_issuer), cnm) != 0)
        continue;
    if (X509_check_akid(crl_issuer, crl->akid) == X509_V_OK) {
        *pissuer = crl_issuer;
        *pcrl_score |= CRL_SCORE_AKID;
        return;
    }
}
```

I use the command:

``` 
openssl verify -extended_crl -show_chain -crl_check_all -CRLfile pki\crl\crls_chain.22.2.pem -CAfile pki\issuer\issuers_chain.1r2.2.pem -untrusted pki\issuer\issuers_chain.11.12.21.22.pem pki\pem\client.12.pem.cer
```

My certificates and CRLs are correct. The situation can change when I rearrange the certificates in the untrusted file. Then the x509_st->skid field will be initialized.

I wrote the following code to check the initialization of x509_st->skid after loading the file. The result is that this field is null.  

```
std::shared_ptr<BIO> bio(BIO_new(BIO_s_file()), ::BIO_free);

if (!BIO_read_filename(bio.get(), file.c_str()))
	BOOST_LOG_TRIVIAL(fatal) << "Error BIO read filename: " << file;

std::shared_ptr<X509> pCert(PEM_read_bio_X509(bio.get(), NULL, NULL, NULL), ::X509_free);

STACK_OF(X509*) st_x509_untrusted = sk_X509_new_null();

sk_X509_push(st_x509_untrusted, cert_root1r2.get());

for (auto i = 0; i < sk_X509_num(st_x509_untrusted); i++) {
	auto crl_issuer = sk_X509_value(st_x509_untrusted, i);
	auto nn = X509_get_subject_name(crl_issuer);
}
```

At what point is the x509_st->skid field initialized?

**Comment by slontis** [comment, marked as answer]
I am not an expert on the X509 code, but doing a code search indicates that it is set via a internal call to 

ossl_x509v3_cache_extensions() (which is not called until it is needed)..
X509_check_purpose() calls this function.. and the check_purpose function is called internally by many functions including getters such as
X509_get0_subject_key_id() and X509_get0_authority_key_id().

**Reply by SergeyKiwi** [reply]
You are right. So I have to call this function before adding the certificate to the store. This looks like a temporary solution, but it works. Thank you for your reply

**Reply by slontis** [reply]
I think this is the only solution, the code is doing lazy evaluation (meaning that it does not do the calculation unless you ask it to).

**Reply by SergeyKiwi** [reply]
Ok, but it was necessary before checking CRL. Maybe OpenSSL should do it before checking.

```
STACK_OF(X509*) st_x509_untrusted = sk_X509_new_null();

sk_X509_push(st_x509_untrusted, cert_root1r2.get());
```

And then 

```
X509_check_akid(crl_issuer, crl->akid) == X509_V_OK
```

Where crl_issuer->skid is NULL. Why didn't he make the calculation earlier when it was necessary? How do I know when I have to do it myself?
